perf: Reduce allocations and printfs in G-function cache.

thetorpedodog opened the pull request:
Previously, the G-function cache used repeated fmt.Sprintf("%v", x) and string concatenation to generate its cache key. This resulted in lots of unnecessary allocations, both in the process of Sprintf and in an O(n²) string building loop. This change allocates everything in one go and relies on the fact that we already know everything is strings.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/casbin/casbin?pullRequest=958) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/casbin/casbin?pullRequest=958) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/casbin/casbin?pullRequest=958) it.</sub>

thetorpedodog commented:
…actually, upon further consideration, we don’t need to concatenate strings at all; we can just use a composite struct key. This further reduces the number of allocations we need to make!

hsluoyz commented:
@tangyang9464 @abichinger @Abingcbc plz review

Abingcbc reviewed (commented):
LGTM

hsluoyz commented:
@thetorpedodog can you paste the benchmark diff here? https://casbin.org/docs/en/benchmark

thetorpedodog commented:
Before:
```
goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkCachedRaw-4                                            76673332                15.70 ns/op            0 B/op          0 allocs/op
BenchmarkCachedBasicModel-4                                      8203942               148.0 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModel-4                                       8185062               146.4 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelSmall-4                                  7929399               150.3 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelMedium-4                                 7978339               150.0 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelLarge-4                                  8778506               133.1 ns/op            96 B/op          3 allocs/op
BenchmarkCachedRBACModelWithResourceRoles-4                      8018532               151.4 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelWithDomains-4                            7484630               160.2 ns/op           120 B/op          4 allocs/op
BenchmarkCachedABACModel-4                                        241495              5271 ns/op            2722 B/op         47 allocs/op
BenchmarkCachedKeyMatchModel-4                                   7015843               166.5 ns/op           152 B/op          4 allocs/op
BenchmarkCachedRBACModelWithDeny-4                               7834410               152.2 ns/op           104 B/op          4 allocs/op
BenchmarkCachedPriorityModel-4                                   8040786               150.7 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelMediumParallel-4                        15450997                74.62 ns/op          105 B/op          4 allocs/op
BenchmarkHasPolicySmall-4                                        2353174               501.3 ns/op           182 B/op          8 allocs/op
BenchmarkHasPolicyMedium-4                                       2276661               538.1 ns/op           189 B/op          8 allocs/op
BenchmarkHasPolicyLarge-4                                        2174547               549.2 ns/op           197 B/op          9 allocs/op
BenchmarkAddPolicySmall-4                                        2249992               534.9 ns/op           184 B/op          8 allocs/op
BenchmarkAddPolicyMedium-4                                       1968679               592.2 ns/op           205 B/op          9 allocs/op
BenchmarkAddPolicyLarge-4                                        1000000              1056 ns/op             506 B/op         11 allocs/op
BenchmarkRemovePolicySmall-4                                     2341110               512.8 ns/op           198 B/op          9 allocs/op
BenchmarkRemovePolicyMedium-4                                    1966756               579.5 ns/op           209 B/op          9 allocs/op
BenchmarkRemovePolicyLarge-4                                     1000000              1064 ns/op             324 B/op         15 allocs/op
BenchmarkRaw-4                                                  76758178                15.68 ns/op            0 B/op          0 allocs/op
BenchmarkBasicModel-4                                             128106              9364 ns/op            4915 B/op         90 allocs/op
BenchmarkRBACModel-4                                               88168             13060 ns/op            6086 B/op        128 allocs/op
BenchmarkRBACModelSmall-4                                           9385            123020 ns/op           44905 B/op       1400 allocs/op
BenchmarkRBACModelMedium-4                                          1065           1121964 ns/op          455403 B/op      13126 allocs/op
BenchmarkRBACModelLarge-4                                            100          11545071 ns/op         4399639 B/op     130265 allocs/op
BenchmarkRBACModelWithResourceRoles-4                              91244             12739 ns/op            6801 B/op        118 allocs/op
BenchmarkRBACModelWithDomains-4                                    77934             15269 ns/op            8102 B/op        144 allocs/op
BenchmarkABACModel-4                                              241245              5003 ns/op            2713 B/op         46 allocs/op
BenchmarkKeyMatchModel-4                                           85480             14154 ns/op            8131 B/op        122 allocs/op
BenchmarkRBACModelWithDeny-4                                       81309             14779 ns/op            6543 B/op        148 allocs/op
BenchmarkPriorityModel-4                                          110157             10901 ns/op            5562 B/op        103 allocs/op
BenchmarkRBACModelWithDomainPatternLarge-4                         35865             32766 ns/op           23102 B/op        280 allocs/op
BenchmarkRoleManagerSmall-4                                        29073             42050 ns/op           13539 B/op        498 allocs/op
BenchmarkRoleManagerMedium-4                                        2758            431509 ns/op          141912 B/op       5742 allocs/op
BenchmarkRoleManagerLarge-4                                          264           4426323 ns/op         1509933 B/op      59742 allocs/op
BenchmarkBuildRoleLinksWithPatternLarge-4                              1        1162629378 ns/op        688769208 B/op  10085340 allocs/op
BenchmarkBuildRoleLinksWithDomainPatternLarge-4                      168           7204872 ns/op         3781946 B/op      89047 allocs/op
BenchmarkBuildRoleLinksWithPatternAndDomainPatternLarge-4              1        1175799088 ns/op        690574704 B/op  10120305 allocs/op
BenchmarkHasLinkWithPatternLarge-4                                 68646             17274 ns/op            4904 B/op        154 allocs/op
BenchmarkHasLinkWithDomainPatternLarge-4                         1000000              1197 ns/op             528 B/op         13 allocs/op
BenchmarkHasLinkWithPatternAndDomainPatternLarge-4                 69610             17136 ns/op            4958 B/op        151 allocs/op
PASS
ok      github.com/casbin/casbin/v2     76.995s
```

After:
```
goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkCachedRaw-4                                            76492984                15.80 ns/op            0 B/op          0 allocs/op
BenchmarkCachedBasicModel-4                                      7648494               154.5 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModel-4                                       7803536               154.3 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelSmall-4                                  7738653               154.2 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelMedium-4                                 7957747               150.8 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelLarge-4                                  8025734               134.6 ns/op            96 B/op          3 allocs/op
BenchmarkCachedRBACModelWithResourceRoles-4                      7495996               156.4 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelWithDomains-4                            7333226               163.9 ns/op           120 B/op          4 allocs/op
BenchmarkCachedABACModel-4                                        237619              5033 ns/op            2729 B/op         47 allocs/op
BenchmarkCachedKeyMatchModel-4                                   6933344               172.3 ns/op           152 B/op          4 allocs/op
BenchmarkCachedRBACModelWithDeny-4                               7765857               157.7 ns/op           104 B/op          4 allocs/op
BenchmarkCachedPriorityModel-4                                   7734268               157.1 ns/op           104 B/op          4 allocs/op
BenchmarkCachedRBACModelMediumParallel-4                        14936707                77.31 ns/op          105 B/op          4 allocs/op
BenchmarkHasPolicySmall-4                                        2330184               522.1 ns/op           182 B/op          8 allocs/op
BenchmarkHasPolicyMedium-4                                       2220385               549.2 ns/op           190 B/op          8 allocs/op
BenchmarkHasPolicyLarge-4                                        2119506               568.4 ns/op           197 B/op          9 allocs/op
BenchmarkAddPolicySmall-4                                        2256610               526.0 ns/op           184 B/op          8 allocs/op
BenchmarkAddPolicyMedium-4                                       1987681               591.9 ns/op           205 B/op          9 allocs/op
BenchmarkAddPolicyLarge-4                                        1000000              1088 ns/op             506 B/op         11 allocs/op
BenchmarkRemovePolicySmall-4                                     2235127               522.5 ns/op           198 B/op          9 allocs/op
BenchmarkRemovePolicyMedium-4                                    1924584               594.1 ns/op           209 B/op          9 allocs/op
BenchmarkRemovePolicyLarge-4                                      992588              1138 ns/op             322 B/op         15 allocs/op
BenchmarkRaw-4                                                  75003219                15.91 ns/op            0 B/op          0 allocs/op
BenchmarkBasicModel-4                                             125618              9525 ns/op            4945 B/op         90 allocs/op
BenchmarkRBACModel-4                                               92558             13187 ns/op            6332 B/op        116 allocs/op
BenchmarkRBACModelSmall-4                                          10000            112212 ns/op           53250 B/op       1000 allocs/op
BenchmarkRBACModelMedium-4                                          1113           1053600 ns/op          581872 B/op       9134 allocs/op
BenchmarkRBACModelLarge-4                                            100          10484457 ns/op         5177710 B/op      90380 allocs/op
BenchmarkRBACModelWithResourceRoles-4                              92574             12694 ns/op            7388 B/op        110 allocs/op
BenchmarkRBACModelWithDomains-4                                    79123             15507 ns/op            8374 B/op        138 allocs/op
BenchmarkABACModel-4                                              237844              5047 ns/op            2716 B/op         46 allocs/op
BenchmarkKeyMatchModel-4                                           82950             14393 ns/op            8153 B/op        122 allocs/op
BenchmarkRBACModelWithDeny-4                                       84972             14448 ns/op            6692 B/op        128 allocs/op
BenchmarkPriorityModel-4                                          109629             10966 ns/op            5849 B/op         99 allocs/op
BenchmarkRBACModelWithDomainPatternLarge-4                         35719             33435 ns/op           23269 B/op        274 allocs/op
BenchmarkRoleManagerSmall-4                                        28006             43793 ns/op           13540 B/op        498 allocs/op
BenchmarkRoleManagerMedium-4                                        2931            433114 ns/op          141913 B/op       5742 allocs/op
BenchmarkRoleManagerLarge-4                                          256           4487358 ns/op         1509928 B/op      59742 allocs/op
BenchmarkBuildRoleLinksWithPatternLarge-4                              1        1172280168 ns/op        689151360 B/op  10085486 allocs/op
BenchmarkBuildRoleLinksWithDomainPatternLarge-4                      160           7283604 ns/op         3786323 B/op      89050 allocs/op
BenchmarkBuildRoleLinksWithPatternAndDomainPatternLarge-4              1        1173982869 ns/op        690965336 B/op  10120419 allocs/op
BenchmarkHasLinkWithPatternLarge-4                                 73510             16953 ns/op            4883 B/op        153 allocs/op
BenchmarkHasLinkWithDomainPatternLarge-4                          965545              1267 ns/op             528 B/op         13 allocs/op
BenchmarkHasLinkWithPatternAndDomainPatternLarge-4                 64540             18024 ns/op            5312 B/op        162 allocs/op
PASS
ok      github.com/casbin/casbin/v2     76.942s
```

The major changes you can see are in the `BenchmarkRBACModel...` benchmarks, where we've gone:

<table>
<tr><th>Benchmark</th><th colspan=2>Before</th><th colspan=2>After</th></tr>
<tr><th>BenchmarkRBACModel-4</th><td>    6069 B/op</td><td>	     128 allocs/op</td><td>    6275 B/op</td><td>	     116 allocs/op</td></tr>
<tr><th>BenchmarkRBACModelSmall-4</th><td>   45013 B/op</td><td>	    1400 allocs/op</td><td>   53238 B/op</td><td>	    1000 allocs/op</td></tr>
<tr><th>BenchmarkRBACModelMedium-4</th><td>  458853 B/op</td><td>	   13127 allocs/op</td><td>  584824 B/op</td><td>	    9134 allocs/op</td></tr>
<tr><th>BenchmarkRBACModelLarge-4</th><td> 4399901 B/op</td><td>	  130263 allocs/op</td><td> 5180076 B/op</td><td>	   90378 allocs/op</td></tr>
<tr><th>BenchmarkRBACModelWithResourceRoles-4</th><td>    6778 B/op</td><td>	     118 allocs/op</td><td>    7346 B/op</td><td>	     110 allocs/op</td></tr>
<tr><th>BenchmarkRBACModelWithDomains-4</th><td>    8084 B/op</td><td>	     144 allocs/op</td><td>    8328 B/op</td><td>	     138 allocs/op</td></tr>
</table>
```

We have a little more memory being allocated in total, but with *significantly* fewer allocations.

I’m not sure how reflective this microbenchmark is of reality, because in one of our systems, a *significant* portion of the time is spent in `Sprintf`, and this will eliminate it entirely.

thetorpedodog commented:
We’ve been running this fix in a staging environment. Without getting into exact numbers, `GenerateGFunction/func1` has gone from accounting for **50%** of the CPU time used in `govaluate.EvaluableExpression.Eval`:

![Eval: 164.73 s; GenerateGFunction/func1: 87.73 s](https://user-images.githubusercontent.com/6622003/154778319-e9a86057-23c6-4545-9798-25cbfa8ff40d.png)

to just **20%** or so:

![Eval: 16.59 s; GenerateGFunction/func1: 3.08 s](https://user-images.githubusercontent.com/6622003/154778398-e7447788-d672-46df-a0fc-6c73373c841e.png)

These are different profiles for different amount of time in different scenarios, so the absolute times and percentages are not comparable, but the workload going into Casbin is similar, so you *can* compare the relative proportion of time spent in `Eval` vs. `GenerateGFunction/func1`. Most importantly the external-facing latency of enforcement calls has **significantly** improved.

hsluoyz commented:
@leeqvip @nodece plz review

leeqvip commented:
@thetorpedodog  Hi, your performance analysis is great, fmt.Sprintf("%v", x) is really bad, so we just need to avoid fmt.Sprintf("%v", x), use `stringify(arg)` instead of it , then use `strings.Builder` to concatenate `args` as key. 

```go

func stringify(it interface{}) string {
	switch it := it.(type) {
	case string:
		return it
	case fmt.Stringer:
		return it.String()
	default:
		return fmt.Sprint(it)
	}
}

func AFunction() func(args ...interface{}) (interface{}, error) {
	type keyType struct {
		name1     string
		name2     string
		domain    string
		hasDomain bool
	}
	memorized := map[keyType]bool{}

	return func(args ...interface{}) (interface{}, error) {
		name1 := args[0].(string)
		name2 := args[1].(string)
		key := keyType{name1, name2, "", false}
		if 2 < len(args) {
			key.domain = stringify(args[2])
			key.hasDomain = true
		}

		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func BFunction() func(args ...interface{}) (interface{}, error) {
	memorized := map[string]bool{}
	return func(args ...interface{}) (interface{}, error) {
		key := ""
		for _, arg := range args {
			key += ";" + stringify(arg)
		}

		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func CFunction() func(args ...interface{}) (interface{}, error) {
	memorized := map[string]bool{}
	var builder strings.Builder
	return func(args ...interface{}) (interface{}, error) {
		builder.Reset()
		for _, arg := range args {
			builder.WriteString(";")
			builder.WriteString(stringify(arg))
		}

		v, found := memorized[builder.String()]
		if found {
			return v, nil
		}
		memorized[builder.String()] = true
		return nil, nil
	}
}

func BenchmarkAFunction(b *testing.B) {
	b.ResetTimer()
	f := BFunction()
	for i := 0; i < b.N; i++ {
		f("user1", "role1", "domo1")
	}
}

func BenchmarkBFunction(b *testing.B) {
	b.ResetTimer()
	f := BFunction()
	for i := 0; i < b.N; i++ {
		f("user1", "role1", "domo1")
	}
}

func BenchmarkCFunction(b *testing.B) {
	b.ResetTimer()
	f := CFunction()
	for i := 0; i < b.N; i++ {
		f("user1", "role1", "domo1")
	}
}

```
```
cpu: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz
BenchmarkAFunction-4   	 4989764	       239.7 ns/op	      96 B/op	       4 allocs/op
BenchmarkBFunction-4   	 5091241	       230.0 ns/op	      96 B/op	       4 allocs/op
BenchmarkCFunction-4   	 6204579	       197.1 ns/op	     104 B/op	       4 allocs/op
```

thetorpedodog commented:
If you want to read this entire comment feel free, but the tl;dr is:

- In this microbenchmark, completely independent of all other things, the `map[keyType]bool` seems to be fastest, and performs 0 allocations.
- In a larger benchmark, where we include the rest of the RBAC machinery, `map[string]bool` seems to win, when combined with pre-allocating a buffer with `builder.Grow`.

I’ve updated this change to pre-allocate a string buffer and build the key therein.

Here’s a long long comment which details the path I went down to get here:

---

The existing `CFunction` above is not thread-safe; if two calls are made at the same time, they will use the same `bytes.Buffer` and overwrite each other.

On the **assumption** that all the values passed to the generated G-function are going to be `string`s (based on the fact that all the datastores exclusively load and serialize strings), I replaced `stringify` with just `.(string)`. **Is this assumption correct?**

Here’s the microbenchmark (which, I’m shocked to see **0** allocations):

```
Running tool: /usr/local/bin/go test -benchmem -run=^$ -bench ^BenchmarkBenchmark$ github.com/casbin/casbin/v2

goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkBenchmark/useKeyType-4         	25462600	        44.06 ns/op	       0 B/op	       0 allocs/op
BenchmarkBenchmark/useStrArrays-4       	13486360	        92.11 ns/op	      40 B/op	       1 allocs/op
BenchmarkBenchmark/useByteSlice-4       	16066762	        75.07 ns/op	      98 B/op	       2 allocs/op
BenchmarkBenchmark/useBuilderAlloc-4    	20858498	        58.06 ns/op	      49 B/op	       1 allocs/op
BenchmarkBenchmark/useBuilder-4         	11623821	       104.6 ns/op	     120 B/op	       3 allocs/op
PASS
ok  	github.com/casbin/casbin/v2	6.389s
```

So the winner in this microbenchmark is `useKeyType` (with `useBuilderAlloc` second). [Compiler Explorer for useKeyType and useBuilderAlloc](https://godbolt.org/z/s8b6sf69T)

```go
package casbin

import (
	"strings"
	"testing"
)

func useKeyType() func(args ...interface{}) (interface{}, error) {
	type keyType struct {
		name1     string
		name2     string
		domain    string
		hasDomain bool
	}
	memorized := map[keyType]bool{}

	return func(args ...interface{}) (interface{}, error) {
		name1 := args[0].(string)
		name2 := args[1].(string)
		key := keyType{name1, name2, "", false}
		if 2 < len(args) {
			key.domain = args[2].(string)
			key.hasDomain = true
		}

		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func useStrArrays() func(args ...interface{}) (interface{}, error) {
	memorized := map[interface{}]bool{}

	return func(args ...interface{}) (interface{}, error) {
		name1 := args[0].(string)
		name2 := args[1].(string)
		var key interface{}
		if len(args) == 2 {
			key = [...]string{name1, name2}
		} else {
			key = [...]string{name1, name2, args[2].(string)}
		}

		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func useByteSlice() func(args ...interface{}) (interface{}, error) {
	memorized := map[string]bool{}
	return func(args ...interface{}) (interface{}, error) {
		total := len(args)
		for _, arg := range args {
			total += len(arg.(string))
		}
		builder := make([]byte, total)
		soFar := 0
		for _, arg := range args {
			astr := arg.(string)
			copy(builder[soFar:], astr)
			soFar += len(astr) + 1
		}

		key := string(builder)
		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func useBuilderAlloc() func(args ...interface{}) (interface{}, error) {
	memorized := map[string]bool{}
	return func(args ...interface{}) (interface{}, error) {
		total := len(args)
		for _, a := range args {
			aStr := a.(string)
			total += len(aStr)
		}
		builder := strings.Builder{}
		builder.Grow(total)
		for _, arg := range args {
			builder.WriteByte(0)
			builder.WriteString(arg.(string))
		}

		key := builder.String()
		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func useBuilder() func(args ...interface{}) (interface{}, error) {
	memorized := map[string]bool{}
	return func(args ...interface{}) (interface{}, error) {
		builder := strings.Builder{}
		for _, arg := range args {
			builder.WriteByte(0)
			builder.WriteString(arg.(string))
		}

		key := builder.String()
		v, found := memorized[key]
		if found {
			return v, nil
		}
		memorized[key] = true
		return nil, nil
	}
}

func BenchmarkBenchmark(b *testing.B) {
	inputs := [][]interface{}{
		{"user", "role"},
		{"ruse", "roll", "frob"},
		{"biggeruser", "biggerrole"},
		{"longeruser", "longerrole", "mcfrobnicate"},
		{"76765613-7dc4-4f88-a06b-7fcbc576ae14", "fa955e85-52b7-492f-a535-8087213ae00b"},
		{"b9fe3e4e-83e0-4604-9b0a-1ac9367a763e", "7a830073-4752-4a4c-ba9e-77a0c790a7d2", "c7c20a86-08e1-4d27-b4f7-76c52aeb753f"},
	}
	fns := []struct {
		name string
		fn   func() func(args ...interface{}) (interface{}, error)
	}{
		{"useKeyType", useKeyType},
		{"useStrArrays", useStrArrays},
		{"useByteSlice", useByteSlice},
		{"useBuilderAlloc", useBuilderAlloc},
		{"useBuilder", useBuilder},
	}
	for _, f := range fns {
		f := f
		fn := f.fn()
		b.Run(f.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				fn(inputs[i%len(inputs)]...)
			}
		})
	}
}
```

And here’s the full-scale benchmark with `useKeyType`:

```
Running tool: /usr/local/bin/go test -benchmem -run=^$ -coverprofile=/tmp/vscode-go4yzFhy/go-code-cover -bench . github.com/casbin/casbin/v2

goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkCachedRaw-4                                        	75171178	        15.80 ns/op	       0 B/op	       0 allocs/op
BenchmarkCachedBasicModel-4                                 	 8538148	       142.5 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModel-4                                  	 8556070	       141.8 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelSmall-4                             	 8155144	       146.8 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelMedium-4                            	 7542036	       157.3 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelLarge-4                             	 7137733	       141.2 ns/op	      96 B/op	       3 allocs/op
BenchmarkCachedRBACModelWithResourceRoles-4                 	 8482335	       141.5 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelWithDomains-4                       	 8200813	       147.6 ns/op	     120 B/op	       4 allocs/op
BenchmarkCachedABACModel-4                                  	  262357	      4578 ns/op	    2708 B/op	      47 allocs/op
BenchmarkCachedKeyMatchModel-4                              	 8019571	       152.7 ns/op	     152 B/op	       4 allocs/op
BenchmarkCachedRBACModelWithDeny-4                          	 8460489	       142.1 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedPriorityModel-4                              	 8420376	       143.4 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelMediumParallel-4                    	 8425383	       135.0 ns/op	     106 B/op	       4 allocs/op
BenchmarkHasPolicySmall-4                                   	 2458981	       486.2 ns/op	     182 B/op	       8 allocs/op
BenchmarkHasPolicyMedium-4                                  	 2277754	       515.5 ns/op	     189 B/op	       8 allocs/op
BenchmarkHasPolicyLarge-4                                   	 2077363	       566.7 ns/op	     197 B/op	       9 allocs/op
BenchmarkAddPolicySmall-4                                   	 2420012	       500.9 ns/op	     184 B/op	       8 allocs/op
BenchmarkAddPolicyMedium-4                                  	 2019922	       606.3 ns/op	     205 B/op	       9 allocs/op
BenchmarkAddPolicyLarge-4                                   	 1000000	      1060 ns/op	     506 B/op	      11 allocs/op
BenchmarkRemovePolicySmall-4                                	 2478364	       498.3 ns/op	     198 B/op	       9 allocs/op
BenchmarkRemovePolicyMedium-4                               	 2106568	       546.4 ns/op	     208 B/op	       9 allocs/op
BenchmarkRemovePolicyLarge-4                                	  991983	      1107 ns/op	     327 B/op	      15 allocs/op
BenchmarkRaw-4                                              	76089723	        15.74 ns/op	       0 B/op	       0 allocs/op
BenchmarkBasicModel-4                                       	  139082	      8609 ns/op	    4897 B/op	      90 allocs/op
BenchmarkRBACModel-4                                        	  100635	     11925 ns/op	    6275 B/op	     116 allocs/op
BenchmarkRBACModelSmall-4                                   	   10000	    106023 ns/op	   53276 B/op	    1000 allocs/op
BenchmarkRBACModelMedium-4                                  	    1118	   1063593 ns/op	  584406 B/op	    9135 allocs/op
BenchmarkRBACModelLarge-4                                   	     100	  10719946 ns/op	 5177649 B/op	   90380 allocs/op
BenchmarkRBACModelWithResourceRoles-4                       	  105168	     11366 ns/op	    7345 B/op	     110 allocs/op
BenchmarkRBACModelWithDomains-4                             	   85749	     13919 ns/op	    8328 B/op	     138 allocs/op
BenchmarkABACModel-4                                        	  263547	      4540 ns/op	    2702 B/op	      46 allocs/op
BenchmarkKeyMatchModel-4                                    	   91680	     12951 ns/op	    8107 B/op	     122 allocs/op
BenchmarkRBACModelWithDeny-4                                	   91507	     13129 ns/op	    6651 B/op	     128 allocs/op
BenchmarkPriorityModel-4                                    	  119842	      9959 ns/op	    5816 B/op	      99 allocs/op
BenchmarkRBACModelWithDomainPatternLarge-4                  	   35628	     34019 ns/op	   23456 B/op	     274 allocs/op
BenchmarkRoleManagerSmall-4                                 	   30298	     40285 ns/op	   13538 B/op	     498 allocs/op
BenchmarkRoleManagerMedium-4                                	    2858	    441686 ns/op	  141932 B/op	    5742 allocs/op
BenchmarkRoleManagerLarge-4                                 	     260	   4567155 ns/op	 1509935 B/op	   59742 allocs/op
BenchmarkBuildRoleLinksWithPatternLarge-4                   	       1	1187175771 ns/op	689813512 B/op	10085667 allocs/op
BenchmarkBuildRoleLinksWithDomainPatternLarge-4             	     163	   7341369 ns/op	 3792799 B/op	   89052 allocs/op
BenchmarkBuildRoleLinksWithPatternAndDomainPatternLarge-4   	       1	1259833072 ns/op	691739896 B/op	10120553 allocs/op
BenchmarkHasLinkWithPatternLarge-4                          	   71545	     16383 ns/op	    4753 B/op	     149 allocs/op
BenchmarkHasLinkWithDomainPatternLarge-4                    	 1000000	      1150 ns/op	     528 B/op	      13 allocs/op
BenchmarkHasLinkWithPatternAndDomainPatternLarge-4          	   64680	     17695 ns/op	    5276 B/op	     161 allocs/op
PASS
coverage: 20.4% of statements
ok  	github.com/casbin/casbin/v2	81.699s
```

versus the full-scale benchmark with `useBuilderAlloc`:

```
Running tool: /usr/local/bin/go test -benchmem -run=^$ -coverprofile=/tmp/vscode-go4yzFhy/go-code-cover -bench . github.com/casbin/casbin/v2

goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkCachedRaw-4                                        	75243138	        15.51 ns/op	       0 B/op	       0 allocs/op
BenchmarkCachedBasicModel-4                                 	 8611347	       139.7 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModel-4                                  	 8722393	       140.7 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelSmall-4                             	 8102049	       147.2 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelMedium-4                            	 7583370	       157.9 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelLarge-4                             	 7825438	       135.8 ns/op	      96 B/op	       3 allocs/op
BenchmarkCachedRBACModelWithResourceRoles-4                 	 8487403	       141.0 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelWithDomains-4                       	 8087373	       149.2 ns/op	     120 B/op	       4 allocs/op
BenchmarkCachedABACModel-4                                  	  261855	      4545 ns/op	    2707 B/op	      47 allocs/op
BenchmarkCachedKeyMatchModel-4                              	 7832830	       153.1 ns/op	     152 B/op	       4 allocs/op
BenchmarkCachedRBACModelWithDeny-4                          	 8571753	       142.2 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedPriorityModel-4                              	 8451496	       141.8 ns/op	     104 B/op	       4 allocs/op
BenchmarkCachedRBACModelMediumParallel-4                    	 8948799	       136.3 ns/op	     105 B/op	       4 allocs/op
BenchmarkHasPolicySmall-4                                   	 2429600	       486.2 ns/op	     182 B/op	       8 allocs/op
BenchmarkHasPolicyMedium-4                                  	 2204013	       515.9 ns/op	     189 B/op	       8 allocs/op
BenchmarkHasPolicyLarge-4                                   	 2015287	       580.3 ns/op	     197 B/op	       9 allocs/op
BenchmarkAddPolicySmall-4                                   	 2351162	       502.4 ns/op	     184 B/op	       8 allocs/op
BenchmarkAddPolicyMedium-4                                  	 1997564	       611.6 ns/op	     205 B/op	       9 allocs/op
BenchmarkAddPolicyLarge-4                                   	 1000000	      1035 ns/op	     506 B/op	      11 allocs/op
BenchmarkRemovePolicySmall-4                                	 2413495	       488.5 ns/op	     198 B/op	       9 allocs/op
BenchmarkRemovePolicyMedium-4                               	 2060479	       563.7 ns/op	     208 B/op	       9 allocs/op
BenchmarkRemovePolicyLarge-4                                	  980253	      1098 ns/op	     323 B/op	      15 allocs/op
BenchmarkRaw-4                                              	78002120	        15.75 ns/op	       0 B/op	       0 allocs/op
BenchmarkBasicModel-4                                       	  135512	      8668 ns/op	    4902 B/op	      90 allocs/op
BenchmarkRBACModel-4                                        	  105147	     11727 ns/op	    5997 B/op	     119 allocs/op
BenchmarkRBACModelSmall-4                                   	   10000	    101017 ns/op	   42703 B/op	    1100 allocs/op
BenchmarkRBACModelMedium-4                                  	    1232	    985505 ns/op	  425797 B/op	   10126 allocs/op
BenchmarkRBACModelLarge-4                                   	     122	   9574341 ns/op	 3929278 B/op	  100263 allocs/op
BenchmarkRBACModelWithResourceRoles-4                       	  105784	     11248 ns/op	    6729 B/op	     112 allocs/op
BenchmarkRBACModelWithDomains-4                             	   85576	     13971 ns/op	    8038 B/op	     139 allocs/op
BenchmarkABACModel-4                                        	  267992	      4556 ns/op	    2702 B/op	      46 allocs/op
BenchmarkKeyMatchModel-4                                    	   91910	     12973 ns/op	    8107 B/op	     122 allocs/op
BenchmarkRBACModelWithDeny-4                                	   93828	     13044 ns/op	    6405 B/op	     133 allocs/op
BenchmarkPriorityModel-4                                    	  121978	      9809 ns/op	    5512 B/op	     100 allocs/op
BenchmarkRBACModelWithDomainPatternLarge-4                  	   34785	     34282 ns/op	   23103 B/op	     275 allocs/op
BenchmarkRoleManagerSmall-4                                 	   29426	     40751 ns/op	   13538 B/op	     498 allocs/op
BenchmarkRoleManagerMedium-4                                	    2695	    451755 ns/op	  141932 B/op	    5742 allocs/op
BenchmarkRoleManagerLarge-4                                 	     265	   4598573 ns/op	 1509931 B/op	   59742 allocs/op
BenchmarkBuildRoleLinksWithPatternLarge-4                   	       1	1448104335 ns/op	689856040 B/op	10085696 allocs/op
BenchmarkBuildRoleLinksWithDomainPatternLarge-4             	     154	   7670547 ns/op	 3796406 B/op	   89052 allocs/op
BenchmarkBuildRoleLinksWithPatternAndDomainPatternLarge-4   	       1	1262826285 ns/op	691358624 B/op	10120612 allocs/op
BenchmarkHasLinkWithPatternLarge-4                          	   69225	     16308 ns/op	    4766 B/op	     149 allocs/op
BenchmarkHasLinkWithDomainPatternLarge-4                    	 1000000	      1192 ns/op	     528 B/op	      13 allocs/op
BenchmarkHasLinkWithPatternAndDomainPatternLarge-4          	   67254	     17932 ns/op	    5215 B/op	     159 allocs/op
PASS
coverage: 20.4% of statements
ok  	github.com/casbin/casbin/v2	84.510s
```

**HOWEVER**, when we integrate the entire thing together, the balance changes:

**keyType** cache:

```
Running tool: /usr/local/bin/go test -benchmem -run=^$ -bench ^BenchmarkRBACModelSizes$ github.com/casbin/casbin/v2

goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkRBACModelSizes/small-4         	   14390	     81945 ns/op	   44158 B/op	     797 allocs/op
BenchmarkRBACModelSizes/medium-4        	    1458	    827354 ns/op	  461398 B/op	    7346 allocs/op
BenchmarkRBACModelSizes/large-4         	     138	   8247208 ns/op	 4223322 B/op	   72568 allocs/op
PASS
ok  	github.com/casbin/casbin/v2	5.686s
```

**string** cache:

```
Running tool: /usr/local/bin/go test -benchmem -run=^$ -bench ^BenchmarkRBACModelSizes$ github.com/casbin/casbin/v2

goos: linux
goarch: amd64
pkg: github.com/casbin/casbin/v2
cpu: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
BenchmarkRBACModelSizes/small-4         	   14040	     78377 ns/op	   38775 B/op	     874 allocs/op
BenchmarkRBACModelSizes/medium-4        	    1507	    763615 ns/op	  372557 B/op	    8143 allocs/op
BenchmarkRBACModelSizes/large-4         	     154	   7904457 ns/op	 3525064 B/op	   80631 allocs/op
PASS
ok  	github.com/casbin/casbin/v2	5.556s
```

These seem fairly reproducible. I suspect (though am not certain) that this has something to do with being able to use `runtime.mapaccess2_faststr`, which, as the name suggests, is fast. Not sure why it wouldn’t show up in the microbenchmark, though.

thetorpedodog commented:
Also, please let me know what you would like me to do with e.g. compressing history to merge this. Thanks!

leeqvip commented:
@thetorpedodog  Good job! I approve of your changes . Please merge the commit history.

thetorpedodog commented:
With pleasure, @leeqvip. I also cleaned up my comments while I was at it, though the functional code is identical.

Thanks for the review!

leeqvip commented:
@hsluoyz  Now we can merge this PR into master.
